### ISSUE TITLE
WebServer password check may expose timing attack

### ISSUE THREAD

**Issue by sweis** [guardianproject/haven#66]
This password check in the WebServer will return on the first character that mPassword and inPassword differ on. This may expose a timing attack where someone can recover the password through repeated queries:
https://github.com/guardianproject/haven/blob/master/src/main/java/org/havenapp/main/service/WebServer.java#L56

**Comment by b-meson** [comment]
ohhhh good find @sweis ! I think you're right

**Comment by noncombatant** [comment]
Note that the same problem is present a few lines down:

```
(!mSession.equals(inSid)))
```

Also it's worth investigating whether or not `java.util.UUID.randomUUID` is cryptographically random on your platform (Android in this case).

**Comment by conorsch** [comment]
@sweis Thanks for reporting this, great find. The timing attack you identify definitely warrants a patch. 

Beyond fixing the basic authentication prompt, there's more we can do to frustrate an attacker's attempts to view the event log remotely. In @micahflee's [OnionShare](https://github.com/micahflee/onionshare/wiki), for example, the Onion URL by appending random words to domain, e.g. <something>.onion/word1-word2. So an attacker would already need to know not only the Onion URL, but the randomly generated route within that domain. 

Better yet, Orbot recently added support for Authenticated Onion Services, which are Onion URLs that cannot be resolved unless the Tor client provides an auth cookie. (See the [Tor manual](https://www.torproject.org/docs/tor-manual.html.en), and search for "HidServAuth".) Setting them up requires a bit more overhead for the user, but the extra work provides additional security guarantees that are hard to beat. I'll open a separate issue to discuss implementing authenticated onion services in Haven.

**Comment by conorsch** [comment]
> I'll open a separate issue to discuss implementing authenticated onion services in Haven.

No need—@mrphs already did so, in #41. :metal:

**Comment by thomasorlando** [comment]
@noncombatant I will check the Android implementation of `java.util.UUID.randomUUID`, but even if the standard implementation might be safe, different (and maybe wrong) implementations are still able to expose vulnerabilities, so it should be considered to add a check here.

**Comment by thomasorlando** [comment]
OK, as far as I can judge, `UUID.randomUUID` implementation is safe.  
([Source file](https://github.com/guardianproject/haven/files/1584686/UUID.java.txt))  
  
The random UUID is generated by a `SecureRandom` generator, which is created according to the [Android guidelines](https://developer.android.com/reference/java/security/SecureRandom.html):
```java
// [Line 95]
private static class Holder {
    static final SecureRandom numberGenerator = new SecureRandom();
}
```

The function generating the UUID seems secure to me, too:
```java
public static UUID randomUUID() {
    SecureRandom ng = Holder.numberGenerator;

    byte[] randomBytes = new byte[16];
    ng.nextBytes(randomBytes);
    randomBytes[6]  &= 0x0f;  /* clear version        */
    randomBytes[6]  |= 0x40;  /* set to version 4     */
    randomBytes[8]  &= 0x3f;  /* clear variant        */
    randomBytes[8]  |= 0x80;  /* set to IETF variant  */
    return new UUID(randomBytes);
}
```
  
  
Although the [specification](https://developer.android.com/reference/java/security/SecureRandom.html) states that
> A cryptographically strong random number minimally complies with the statistical random number generator tests specified in [_FIPS 140-2, Security Requirements for Cryptographic Modules_](http://csrc.nist.gov/cryptval/140-2.htm), section 4.9.1,

different implementations of the `SecureRandom` class might not follow this guidelines and therefore _could_ break the security of the randomly generated UUID's; but the implementation itself is secure.

**Comment by n8fr8** [comment]
Any comments on the fix supplied in #88 ? The safeEquals() method seems like it addresses the main points raised here.

**Comment by b-meson** [comment]
@n8fr8 I think that PR looks good.
